docs/proposals/query-blocking.md

```diff
@@ -0,0 +1,110 @@
+---
+title: "XXXX: Query blocking"
+description: "Query blocking"
+draft: true
+---
+
+# XXXX: Query blocking
+
+**Author:** @wilfriedroset
+
+**Date:** 05/2023
+
+**Sponsor(s):** @username of maintainer(s) willing to shepherd this MID
+
+**Type:** Feature
+
+**Status:** Draft
+
+**Related issues/PRs:**
+
+---
+
+## Background
+
+The `query-frontend` is a component that is responsible for query acceleration through splitting, sharding, caching, and other techniques. It is also responsible for enforcing limits on queries such as total query time range.
+
+## Problem Statement
+
+In certain situations, you may not be able to control the queries being sent to your Mimir installation. These queries may be intentionally or unintentionally expensive to run, and they may affect the overall stability or cost of running your service.
+
+Example:
+
+```
+sum(increase(tfe_run_current_count{status=~"applied|planned_and_finished", instance="host1", organization_name=~".*", workspace_name=~".*"}[2674800s]) > 0)
+```
+
+The query above can be harmful as the wildcard regex would match any label values including empty string. The result of such query is not cached by Mimir. As such the read path would be under pressure if the query is repeated by the tenant (e.g: dashboard in auto repeat).
+
+## Goals
+
+- Prevent harmful queries to be processed by Mimir's read path.
+
+## Non-Goals (optional)
+
+- Rewrite queries in a non harmful way.
+- Optimize queries on the fly.
+- Blocked remote read.
+
+## Proposals
+
+### Proposal 0: Do nothing
+
+Current tenants isolation through `shuffle-sharding` are sufficient to keep the read path healthy. Mimir's operators can use limits such as `max_fetched_chunks_per_query`, `max_queriers_per_tenant` or `store_gateway_tenant_shard_size` to limits tenants processing capabilities.
+
+**Pros:**
+
+- Nothing to be done
+
+**Cons:**
+
+- Harmful queries are still processed by Mimir but in a limited part of the cluster
+
+### Proposal 1: Query blocking
+
+Based on the configuration, Mimir's [limitsMiddleware](https://github.com/grafana/mimir/blob/main/pkg/frontend/querymiddleware/limits.go)) runned by the `query-frontend` would be responsible for blocking queries. Operators would be able to block queries either for all tenants or on a per-tenant basis via the _runtime configuration_.
+
+```yaml
+overrides:
+  "tenant-id":
+    blocked_queries:
```

Comment:
A couple of questions:
1. What if we just allow to configure 1 regexp matching the query, to start with an easier solution?
2. Do we really need `hash`? What's the actual use case for that?

Reply:
> What if we just allow to configure 1 regexp matching the query, to start with an easier solution?

Yes we could start with this simpler version. The risk here is to have a hard to read regex

> Do we really need hash? What's the actual use case for that?

Depending on the complexity of the query you want to block the configuration could be rather big. Having the possibility to block a query based on the `hash` makes it simple. More over, as an administrator, I don't need to pay attention to escaping anything if I just block a query based on the `hash` outputed in mimir's logs.

Reply:
> I don't need to pay attention to escaping anything

are YAML multi-line strings an alternative?

Reply:
> are YAML multi-line strings an alternative?

yes this could be a solution. I will take into account your two comments and remove the `hash` and start with 1 regexp per tenant.

Reply:
I think it's reasonable to have both regex and non-regex because of the escaping problem and how many characters in promQL are also special characters in regexps (`[](){},-`) and that promQL itself can contain regexp.

My concern with the hash is that it's difficult/error-prone to track which hash corresponds to which query from the operator point of view. It is also somewhat error-prone to construct a hash - do I need to remove spaces from the query, should I format it beforehand, etc.

Reply:
Hi,

> My concern with the hash is that it's difficult/error-prone to track which hash corresponds to which query from the operator point of view. It is also somewhat error-prone to construct a hash - do I need to remove spaces from the query, should I format it beforehand, etc.

Hash of query could be computed by mimir and displayed with the query inside the log, this way their won't be any mistake possible.

> With or without the hash we have a similar issue when using regex or plain query. If an operator want to block foo{label1='value1', label2='value2'} I believe this person also expected that foo{ label1='value1', label2='value2' } and foo{label2='value2', label1='value1'} are blocked as well as they are the same query with a different formating.

I'm afraid I must disagree with you, even if those two queries should produce the same results, it doesn't mean they will be interacting the same way with mimir and it's dependencies (cache, object storage...), especially if you use regex as labels selector.

Regards,

Reply:
> Hash of query could be computed by mimir and displayed with the query inside the log, this way their won't be any mistake possible.

It is still difficult when reviewing existing configuration. An operator can't know whether they still want to block a query hash. That forces them to track query:hash mappings in comments or some other form, which can be wrong or become outdated.

> if those two queries should produce the same results, it doesn't mean they will be interacting the same way with mimir and it's dependencies

Currently this order matters only for the ingester but IMO it's negligible. I think it's easier for an operator to think of this as the same query. Knowing whether it will be more expensive requires knowledge of the prometheus TSDB querying code and that's also subject to change.

Reply:
Thank you @dimitarvdimitrov for your feedbacks. I've updated the PR to remove the `hash` part and move to *non-goal* blocking similar queries.
WDYT about this update?

Reply:
awesome, thanks for updating it!